cfg_mgmt/btp_service_binding.py

```diff
@@ -0,0 +1,159 @@
+import copy
+import json
+import logging
+import typing
+import requests
+
+import cfg_mgmt
+import cfg_mgmt.model as cmm
+import ci.log
+import ci.util
+import model
+import model.container_registry
+
+
+ci.log.configure_default_logging()
+logger = logging.getLogger(__name__)
+
+
+def _filter_service_bindings(result: dict, binding_name_prefix: str, instance_id: str) -> tuple[list[str], int]:
+    outdated_ids = []
+    next = 1
+    for item in result["items"]:
+        name = item["name"]
+        if name.startswith(binding_name_prefix) and item["service_instance_id"] == instance_id:
+            try:
+                n = int(name[len(binding_name_prefix):])
+                outdated_ids.append(item["id"])
+                if n >= next:
+                    next = n + 1
+            except ValueError:
+                logger.warn("ignored {}".format(name))
+    return outdated_ids, next
+
+
+class sbClient:
+    def __init__(
+        self,
+        sm_url: str,
+        access_token,
+    ):
+        self.sm_url = sm_url
+        self.access_token = access_token
+
+    def delete_service_binding(self, name: str, id: str):
+        headers = {
+            'Authorization': 'Bearer {}'.format(self.access_token),
+        }
+        url = "{}/v1/service_bindings/{}".format(self.sm_url, id)
+        resp = requests.delete(url, headers=headers)
+        if resp.status_code != 200:
+            msg = 'delete_service_binding failed: {} {}'.format(resp.status_code, resp.text)
+            logger.error(msg)
+            raise requests.HTTPError(msg)
+        logger.info("Deleted service binding {} ({})".format(name, id))
+
+    def create_service_binding(self, instance_id: str, binding_name: str) -> tuple[str, dict]:
+        url = "{}/v1/service_bindings".format(self.sm_url)
+        data = {
+            "name": binding_name,
+            "service_instance_id": instance_id,
+        }
+        headers = {
+            'Accept': 'application/json',
+            'Authorization': 'Bearer {}'.format(self.access_token),
+            'Content-Type': 'application/json',
+        }
+        resp = requests.post(url, data=json.dumps(data), headers=headers)
+        if resp.status_code != 201:
+            msg = 'create_service_binding failed: {} {}'.format(resp.status_code, resp.text)
+            logger.error(msg)
+            raise requests.HTTPError(msg)
+        result = resp.json()
+        id = result["id"]
+        logger.info('Creating service binding {} ({})'.format(binding_name, id))
+        return id, result["credentials"]
+
+    def get_service_bindings(self) -> dict:
+        url = "{}/v1/service_bindings".format(self.sm_url)
+        headers = {
+            'Accept': 'application/json',
+            'Authorization': 'Bearer {}'.format(self.access_token),
```

Comment:
since there is a lot of redundancy w.r.t. setting those headers, how about either defining a wrapper method for requests.request|get|post that will patch-in those?

example: https://github.com/gardener/cc-utils/blob/master/oci/client.py#L356-L421 (most relevantly consider at last line; requests.Session is (almost) identical to requests w.r.t. signatures)

Reply:
In general I agree with removing redundancy, but here I don't see that it would improve the code. There would be more code lines and there will be no reuse.